### ISSUE TITLE
Netty event loop may be leaked

### ISSUE THREAD

**Issue by EronWright** [pravega/pravega#4401]
**Problem description**
We've observed that, in certain circumstances, the Pravega client appears to leak some Netty threads.  The existence of which is occasionally observable, e.g. when Flink unloads a job (upon cancellation) including its classloader.   Here's a sampling:

```
2019-11-26 22:33:06,479 WARN  io.netty.channel.epoll.EpollEventLoop                         - Unexpected exception in the selector loop.
java.lang.NoClassDefFoundError: io/netty/util/collection/IntObjectHashMap$2
        at io.netty.util.collection.IntObjectHashMap.values(IntObjectHashMap.java:221)
        at io.netty.channel.epoll.EpollEventLoop.closeAll(EpollEventLoop.java:372)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:339)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
```
And:
```
SEVERE: Exception while executing runnable io.grpc.internal.KeepAliveManager$1@12240d6b
java.lang.NoClassDefFoundError: io/grpc/netty/NettyClientTransport$6
 at io.grpc.netty.NettyClientTransport.shutdownNow(NettyClientTransport.java:302)
 at io.grpc.internal.KeepAliveManager$ClientKeepAlivePinger.onPingTimeout(KeepAliveManager.java:287)
 at io.grpc.internal.KeepAliveManager$1.run(KeepAliveManager.java:63)
 at io.grpc.internal.LogExceptionRunnable.run(LogExceptionRunnable.java:41)
 at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38)
 at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:127)
 at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
 at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
 at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:462)
 at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
 at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
 at java.lang.Thread.run(Thread.java:748)
```

**Problem location**
Here's a few possible locations which could be a problem:
1. `io.pravega.client.stream.impl.ControllerResolverFactory`
2. `io.pravega.client.stream.impl.ControllerImpl` (a channel is leaked if the creds plugin throws an exception ([ref](https://github.com/pravega/pravega/blob/v0.6.0/client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java#L186-L193)))

The latter location seems likely, because we observed this behavior in the course of investigating a DNS issue that causes the Keycloak auth plugin to fail to initialize.

**Comment by ravisharda** [comment]
Here are some observations and analysis from sometime back:

1. The classes that failed to get loaded from classpath, come from via these dependencies. (I had verified this in an experimental Flink app that uses Pravega connector to read/write from Pravega using Pravega Flink connector and runs a  Flink job.) 

  ```
  io/grpc/netty/NettyClientTransport$6: 
pravega-client > pravega-shared-controller-api > grp-netty
 io/netty/util/collection/IntObjectHashMap$2:
pravega-client > pravega-shared-protocol > netty-transport > grpc-common
  ```

2. The jars containing the classes which failed to load above must have been present in the job for sure for the Pravega connector to have worked. However, we still see these errors. There could have been either a thread or an object running beyond a Job's lifetime.

3. These are not the only classes that fail loading. On occassion, we have seen other classes failing to load as well under different circumstances and test runs. A couple of examples were provided in another issue: https://github.com/pravega/flink-connectors/issues/136. Interestingly, in all these cases, the classes were related to gRPC Netty transport. 

4. I scanned through logs from other Longevity runs and found exceptions in loading other classes at different times. Copying some of the instances from other runs in the past under the thread. At the outset, an interesting thing to note is that most of these errors  come in the path of gRPC operations on the client side. 

5. At the outset, the issue doesn't seem to be resulting in any perceivable functional error. 

6. Interestingly, in all cases that I have seen so far, the log entry indicating failure to load the class occurs soon after (2-6 lines in distance from) this log entry: 
  
   ```
   io.pravega.client.stream.impl.ControllerResolverFactory       - Attempting to refresh the controller server endpoints
   ```
